### PR TITLE
Use a separate install.Lock for each dependency

### PR DESCRIPTION
**Pull request by akshita31**
Fixes: https://github.com/OmniSharp/omnisharp-vscode/issues/2478

With this change we will have an install.Lock for each package that is downloaded rather than having a single install.Lock for the runtime dependencies.
This will be a step forward towards the validation of the downloads as listed here: https://github.com/OmniSharp/omnisharp-vscode/issues/2694.

Also, in this PR, I have added an "Id" attribute to the packages so that whatever part of the code needs access to a particular package to verify if the required dependency is installed or not, they could filter the packages from package.json easily, eg, for the case of the debugger packages as I have done in this PR.

cc @NTaylorMullen

### CODE REVIEW

**Comment by rchande** [comment]
How do this work if I upgrade an existing installation? We'll just stop looking at my existing install.lock and be good to go?

File: src/packageManager/AbsolutePathPackage.ts [hunk 1]
```diff
@@ -8,7 +8,8 @@ import { IPackage } from "./IPackage";
 import { AbsolutePath } from "./AbsolutePath";
 
 export class AbsolutePathPackage implements IPackage{
-    constructor(public description: string,
+    constructor(public Id: string,
```
Comment (gregg-miskelly) [inline src/packageManager/AbsolutePathPackage.ts]: nit: should this be cased with a lower case 'id'?

File: src/packageManager/PackageManager.ts [hunk 2]
```diff
@@ -13,24 +13,41 @@ import { EventStream } from '../EventStream';
...
+                    eventStream.post(new InstallationFailure(installationStage, packageError));
+                    throw packageError;
                 }
                 else {
+                    eventStream.post(new InstallationFailure(installationStage, error));
                     throw error;
                 }
             }
+            finally {
+                await deleteInstallFile(pkg.installPath, InstallFileType.Begin);
```
Comment (gregg-miskelly) [inline src/packageManager/PackageManager.ts]: Should we only do this if we created the Begin file? Or do we need to eat exceptions if it doesn't exist?

File: src/coreclr-debug/activate.ts [hunk 3]
```diff
@@ -125,7 +126,14 @@ export async function getAdapterExecutionCommand(platformInfo: PlatformInformati
...
         // install.Lock does not exist, need to wait for packages to finish downloading.
-        let installLock: boolean = await common.installFileExists(common.InstallFileType.Lock);
+        let installLock = false;
+        let debuggerPackage = getRuntimeDependencyPackageWithId("Debugger", packageJSON, platformInfo, extensionPath);
+        if (debuggerPackage && debuggerPackage.installPath)
+        {
+            installLock = await common.installFileExists(debuggerPackage.installPath, common.InstallFileType.Lock);
+        }
+        
+        //todo: check if the debugger install.lock is present
```
Comment (gregg-miskelly) [inline src/coreclr-debug/activate.ts]: Should this comment still be here?

File: src/CSharpExtDownloader.ts [hunk 4]
```diff
@@ -26,30 +25,18 @@ export class CSharpExtDownloader {
 
     public async installRuntimeDependencies(): Promise<boolean> {
```
Comment (gregg-miskelly) [inline src/CSharpExtDownloader.ts]: We are now always calling this method when the C# extension starts right? If so, shouldn't we avoid posting some of these events to the event stream until after we have determined if there are any missing packages?

Comment (akshita31) [reply src/CSharpExtDownloader.ts]: @gregg-miskelly You are right, but with the current code structure it is not possible and we will need some refactoring in order to achieve that. I will do that in a subsequent PR.

**Comment by akshita31** [comment]
@NTaylorMullen Do you have any concerns with this change ?

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen Do you have any concerns with this change ?

Based on what you wrote in the description of the PR this sounds great! Means we wont have busted partial installs of dependencies it sounds like. Any negative implications to this?

**Comment by akshita31** [comment]
@NTaylorMullen I am still unsure of what part fails on the user's systems when they have an improper download/install. With this PR, I think it will help us identify better which of the three dependencies got corrupted as each will have its own install.Lock file that *should* be put only when both the download and install completed without throwing an exception. 

And if for a dependency say OmniSharp, the download failed this time and the install.Lock was not put into the expected location, the next time the user opens vscode, it would detect that it needs to install Omnisharp as the install.Lock is not present.

**Comment by NTaylorMullen** [comment]
Awesome, ya this would be a great addition, no concerns!

**Pull request merged**